src/lib/operators/projection.cpp

```diff
@@ -97,47 +97,109 @@ std::shared_ptr<const Table> Projection::_on_execute() {
     const auto input_chunk = input_table.get_chunk(chunk_id);
     Assert(input_chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
 
-    auto output_segments = Segments{expressions.size()};
+    auto output_segments = Segments{expression_count};
+    auto forwared_count = size_t{0};
 
-    // The ExpressionEvaluator is created once per chunk so that evaluated sub-expressions can be reused across columns.
-    ExpressionEvaluator evaluator(left_input_table(), chunk_id, uncorrelated_subquery_results);
-
-    for (auto column_id = ColumnID{0}; column_id < expressions.size(); ++column_id) {
+    for (auto column_id = ColumnID{0}; column_id < expression_count; ++column_id) {
+      // In this loop, we perform all projections that only forward an input column sequential.
       const auto& expression = expressions[column_id];
-
       if (expression->type == ExpressionType::PQPColumn) {
         // Forward input column if possible
+        forwared_count++;
         const auto& pqp_column_expression = static_cast<const PQPColumnExpression&>(*expression);
         output_segments[column_id] = input_chunk->get_segment(pqp_column_expression.column_id);
-        column_is_nullable[column_id] =
-            column_is_nullable[column_id] || input_table.column_is_nullable(pqp_column_expression.column_id);
-        forwarding_cost += timer.lap();
-      } else {
-        // Newly generated column - the expression needs to be evaluated
-        auto output_segment = evaluator.evaluate_expression_to_segment(*expression);
-        column_is_nullable[column_id] = column_is_nullable[column_id] || output_segment->is_nullable();
-
-        // Storing the result in output_segments means that the vector may contain both ReferenceSegments and
-        // ValueSegments. We deal with this later.
-        output_segments[column_id] = std::move(output_segment);
-        expression_evaluator_cost += timer.lap();
+        column_is_nullable_by_chunk[chunk_id][column_id] =
+          column_is_nullable_by_chunk[chunk_id][column_id] ||
+          input_table.column_is_nullable(pqp_column_expression.column_id);
       }
     }
+    forwarding_cost += timer.lap();
 
+    // Output segments are associated with their chunk. At the moment the segments are only ReferenceSegments.
     output_segments_by_chunk[chunk_id] = std::move(output_segments);
+
+    // All columns are forwarded. We do not need to evaluate newly generated columns.
+    if (forwared_count == expression_count) continue;
+
+    // Defines the job that performs the evaluation if the columns are newly generated.
+    auto perform_projection_evaluation = [
+      this,
+      chunk_id,
+      &uncorrelated_subquery_results,
+      expression_count,
+      &output_segments_by_chunk,
+      &column_is_nullable_by_chunk
+      ] () {
+        ExpressionEvaluator evaluator(left_input_table(), chunk_id, uncorrelated_subquery_results);
```

Comment:
Not your code, but we are slowly moving to [auto-to-stick](https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/):

`auto evaluator = ExpressionEvaluator{left_input_table(), chunk_id, uncorrelated_subquery_results};`